#!/usr/bin/env python3
"""
–¢–µ—Å—Ç –ø–æ–ª–Ω–æ–π —Ü–µ–ø–æ—á–∫–∏ –¥–æ gRPC: –ø—Ä–æ–≤–µ—Ä–∫–∞ —Ñ–æ—Ä–º–∞—Ç–∞ –¥–∞–Ω–Ω—ã—Ö –¥–ª—è –æ—Ç–ø—Ä–∞–≤–∫–∏ –∫–ª–∏–µ–Ω—Ç—É

–ü—Ä–æ–≤–µ—Ä—è–µ—Ç:
1. command_payload –ø—Ä–∞–≤–∏–ª—å–Ω–æ —Ñ–æ—Ä–º–∏—Ä—É–µ—Ç—Å—è
2. –î–∞–Ω–Ω—ã–µ –ø—Ä–∞–≤–∏–ª—å–Ω–æ —Å–µ—Ä–∏–∞–ª–∏–∑—É—é—Ç—Å—è –¥–ª—è gRPC
3. –§–æ—Ä–º–∞—Ç —Å–æ–æ—Ç–≤–µ—Ç—Å—Ç–≤—É–µ—Ç proto –∫–æ–Ω—Ç—Ä–∞–∫—Ç—É
"""

import sys
import os
import json
from pathlib import Path

# –î–æ–±–∞–≤–ª—è–µ–º –ø—É—Ç—å –∫ —Å–µ—Ä–≤–µ—Ä—É
project_root = Path(__file__).parent.parent.parent
sys.path.insert(0, str(project_root / "server"))

from modules.grpc_service import streaming_pb2


def test_command_payload_format():
    """–¢–µ—Å—Ç 1: –§–æ—Ä–º–∞—Ç command_payload –¥–ª—è gRPC"""
    print("\n" + "="*80)
    print("–¢–ï–°–¢ 1: –§–æ—Ä–º–∞—Ç command_payload –¥–ª—è gRPC")
    print("="*80)
    
    # –°–∏–º—É–ª–∏—Ä—É–µ–º command_payload –∏–∑ workflow
    command_payload = {
        'event': 'mcp.command_request',
        'payload': {
            'session_id': 'test_session_123',
            'command': 'open_app',
            'args': {
                'app_name': 'Safari'
            }
        }
    }
    
    # –ü—Ä–æ–≤–µ—Ä—è–µ–º —Ñ–æ—Ä–º–∞—Ç –¥–ª—è –æ—Ç–ø—Ä–∞–≤–∫–∏ —á–µ—Ä–µ–∑ text_chunk —Å –ø—Ä–µ—Ñ–∏–∫—Å–æ–º __MCP__
    try:
        mcp_json = json.dumps(command_payload, ensure_ascii=False)
        mcp_text_chunk = f"__MCP__{mcp_json}"
        
        print(f"\nüìã –§–æ—Ä–º–∞—Ç –¥–∞–Ω–Ω—ã—Ö:")
        print(f"   ‚úÖ JSON –¥–ª–∏–Ω–∞: {len(mcp_json)}")
        print(f"   ‚úÖ –° –ø—Ä–µ—Ñ–∏–∫—Å–æ–º: {len(mcp_text_chunk)}")
        print(f"   ‚úÖ –ü—Ä–µ—Ñ–∏–∫—Å: __MCP__")
        print(f"   ‚úÖ JSON –≤–∞–ª–∏–¥–µ–Ω: {json.loads(mcp_json) is not None}")
        
        # –ü—Ä–æ–≤–µ—Ä—è–µ–º —Å—Ç—Ä—É–∫—Ç—É—Ä—É
        parsed = json.loads(mcp_json)
        payload = parsed.get('payload', {})
        command = payload.get('command')
        args = payload.get('args', {})
        app_name = args.get('app_name') if isinstance(args, dict) else None
        
        if command == 'open_app' and app_name == 'Safari':
            print(f"   ‚úÖ –°—Ç—Ä—É–∫—Ç—É—Ä–∞ –∫–æ—Ä—Ä–µ–∫—Ç–Ω–∞: command={command}, app={app_name}")
            print(f"\n‚úÖ –¢–ï–°–¢ 1 –ü–†–û–ô–î–ï–ù: –§–æ—Ä–º–∞—Ç command_payload –∫–æ—Ä—Ä–µ–∫—Ç–µ–Ω")
            return True
        else:
            print(f"   ‚ùå –°—Ç—Ä—É–∫—Ç—É—Ä–∞ –Ω–µ–∫–æ—Ä—Ä–µ–∫—Ç–Ω–∞: command={command}, app={app_name}")
            return False
            
    except Exception as e:
        print(f"   ‚ùå –û—à–∏–±–∫–∞: {e}")
        return False


def test_action_message_proto():
    """–¢–µ—Å—Ç 2: –§–æ—Ä–º–∞—Ç ActionMessage –¥–ª—è proto (–µ—Å–ª–∏ –∏—Å–ø–æ–ª—å–∑—É–µ—Ç—Å—è)"""
    print("\n" + "="*80)
    print("–¢–ï–°–¢ 2: –§–æ—Ä–º–∞—Ç ActionMessage –¥–ª—è proto")
    print("="*80)
    
    # –°–æ–∑–¥–∞–µ–º ActionMessage —Å–æ–≥–ª–∞—Å–Ω–æ proto
    action_json = json.dumps({
        "type": "open_app",
        "app_name": "Safari"
    })
    
    try:
        action_message = streaming_pb2.ActionMessage(
            action_json=action_json,
            session_id="test_session_123",
            feature_id="F-2025-013-open-app"
        )
        
        print(f"\nüìã ActionMessage:")
        print(f"   ‚úÖ action_json: {action_message.action_json[:50]}...")
        print(f"   ‚úÖ session_id: {action_message.session_id}")
        print(f"   ‚úÖ feature_id: {action_message.feature_id}")
        
        # –ü—Ä–æ–≤–µ—Ä—è–µ–º —Å–æ–¥–µ—Ä–∂–∏–º–æ–µ action_json
        parsed_action = json.loads(action_message.action_json)
        if parsed_action.get('type') == 'open_app' and parsed_action.get('app_name') == 'Safari':
            print(f"   ‚úÖ action_json —Å—Ç—Ä—É–∫—Ç—É—Ä–∞ –∫–æ—Ä—Ä–µ–∫—Ç–Ω–∞")
            print(f"\n‚úÖ –¢–ï–°–¢ 2 –ü–†–û–ô–î–ï–ù: ActionMessage —Ñ–æ—Ä–º–∞—Ç –∫–æ—Ä—Ä–µ–∫—Ç–µ–Ω")
            return True
        else:
            print(f"   ‚ùå action_json —Å—Ç—Ä—É–∫—Ç—É—Ä–∞ –Ω–µ–∫–æ—Ä—Ä–µ–∫—Ç–Ω–∞")
            return False
            
    except Exception as e:
        print(f"   ‚ùå –û—à–∏–±–∫–∞ —Å–æ–∑–¥–∞–Ω–∏—è ActionMessage: {e}")
        return False


def test_stream_response_format():
    """–¢–µ—Å—Ç 3: –§–æ—Ä–º–∞—Ç StreamResponse –¥–ª—è —Ä–∞–∑–Ω—ã—Ö —Ç–∏–ø–æ–≤ –∫–æ–Ω—Ç–µ–Ω—Ç–∞"""
    print("\n" + "="*80)
    print("–¢–ï–°–¢ 3: –§–æ—Ä–º–∞—Ç StreamResponse")
    print("="*80)
    
    test_cases = [
        {
            "name": "Text chunk",
            "create": lambda: streaming_pb2.StreamResponse(text_chunk="Hello!")
        },
        {
            "name": "Audio chunk",
            "create": lambda: streaming_pb2.StreamResponse(
                audio_chunk=streaming_pb2.AudioChunk(
                    audio_data=b"fake_audio",
                    dtype="int16",
                    shape=[1024]
                )
            )
        },
        {
            "name": "Action message",
            "create": lambda: streaming_pb2.StreamResponse(
                action_message=streaming_pb2.ActionMessage(
                    action_json='{"type":"open_app","app_name":"Safari"}',
                    session_id="test_session_123"
                )
            )
        },
        {
            "name": "End message",
            "create": lambda: streaming_pb2.StreamResponse(end_message="Done")
        }
    ]
    
    all_passed = True
    for i, test_case in enumerate(test_cases, 1):
        print(f"\nüìù –¢–µ—Å—Ç 3.{i}: {test_case['name']}")
        try:
            response = test_case['create']()
            
            # –ü—Ä–æ–≤–µ—Ä—è–µ–º, —á—Ç–æ –ø—Ä–∞–≤–∏–ª—å–Ω–æ —É—Å—Ç–∞–Ω–æ–≤–ª–µ–Ω oneof
            which_oneof = response.WhichOneof('content')
            print(f"   ‚úÖ WhichOneof: {which_oneof}")
            
            if which_oneof == test_case['name'].lower().replace(' ', '_'):
                print(f"   ‚úÖ –¢–∏–ø –∫–æ–Ω—Ç–µ–Ω—Ç–∞ –∫–æ—Ä—Ä–µ–∫—Ç–µ–Ω")
            else:
                print(f"   ‚ö†Ô∏è  –¢–∏–ø –∫–æ–Ω—Ç–µ–Ω—Ç–∞: {which_oneof} (–æ–∂–∏–¥–∞–ª—Å—è –¥—Ä—É–≥–æ–π —Ñ–æ—Ä–º–∞—Ç)")
                
        except Exception as e:
            print(f"   ‚ùå –û—à–∏–±–∫–∞: {e}")
            all_passed = False
    
    if all_passed:
        print(f"\n‚úÖ –¢–ï–°–¢ 3 –ü–†–û–ô–î–ï–ù: StreamResponse —Ñ–æ—Ä–º–∞—Ç –∫–æ—Ä—Ä–µ–∫—Ç–µ–Ω")
        return True
    else:
        return False


def test_mcp_text_chunk_format():
    """–¢–µ—Å—Ç 4: –§–æ—Ä–º–∞—Ç __MCP__ text_chunk –¥–ª—è –∫–ª–∏–µ–Ω—Ç–∞"""
    print("\n" + "="*80)
    print("–¢–ï–°–¢ 4: –§–æ—Ä–º–∞—Ç __MCP__ text_chunk")
    print("="*80)
    
    command_payload = {
        'event': 'mcp.command_request',
        'payload': {
            'session_id': 'test_session_456',
            'command': 'open_app',
            'args': {
                'app_name': 'Telegram'
            }
        }
    }
    
    mcp_json = json.dumps(command_payload, ensure_ascii=False)
    mcp_text_chunk = f"__MCP__{mcp_json}"
    
    # –°–∏–º—É–ª–∏—Ä—É–µ–º StreamResponse
    response = streaming_pb2.StreamResponse(text_chunk=mcp_text_chunk)
    
    print(f"\nüìã MCP text_chunk:")
    print(f"   ‚úÖ –î–ª–∏–Ω–∞: {len(mcp_text_chunk)}")
    print(f"   ‚úÖ –ù–∞—á–∏–Ω–∞–µ—Ç—Å—è —Å __MCP__: {mcp_text_chunk.startswith('__MCP__')}")
    print(f"   ‚úÖ JSON –≤–∞–ª–∏–¥–µ–Ω: {json.loads(mcp_text_chunk[7:]) is not None}")
    
    # –ü—Ä–æ–≤–µ—Ä—è–µ–º, —á—Ç–æ –∫–ª–∏–µ–Ω—Ç —Å–º–æ–∂–µ—Ç –∏–∑–≤–ª–µ—á—å
    if mcp_text_chunk.startswith('__MCP__'):
        extracted_json = mcp_text_chunk[7:]  # –£–±–∏—Ä–∞–µ–º –ø—Ä–µ—Ñ–∏–∫—Å
        parsed = json.loads(extracted_json)
        payload = parsed.get('payload', {})
        command = payload.get('command')
        
        if command == 'open_app':
            print(f"   ‚úÖ –ö–æ–º–∞–Ω–¥–∞ –∏–∑–≤–ª–µ—á–µ–Ω–∞: {command}")
            print(f"\n‚úÖ –¢–ï–°–¢ 4 –ü–†–û–ô–î–ï–ù: –§–æ—Ä–º–∞—Ç __MCP__ text_chunk –∫–æ—Ä—Ä–µ–∫—Ç–µ–Ω")
            return True
        else:
            print(f"   ‚ùå –ö–æ–º–∞–Ω–¥–∞ –Ω–µ –∏–∑–≤–ª–µ—á–µ–Ω–∞")
            return False
    else:
        print(f"   ‚ùå –ü—Ä–µ—Ñ–∏–∫—Å –æ—Ç—Å—É—Ç—Å—Ç–≤—É–µ—Ç")
        return False


def main():
    """–ó–∞–ø—É—Å–∫ –≤—Å–µ—Ö —Ç–µ—Å—Ç–æ–≤ gRPC —Ñ–æ—Ä–º–∞—Ç–∞"""
    print("\n" + "="*80)
    print("–¢–ï–°–¢–ò–†–û–í–ê–ù–ò–ï –§–û–†–ú–ê–¢–ê –î–ê–ù–ù–´–• –î–õ–Ø gRPC")
    print("="*80)
    print("\n–ü—Ä–æ–≤–µ—Ä–∫–∞:")
    print("1. –§–æ—Ä–º–∞—Ç command_payload")
    print("2. –§–æ—Ä–º–∞—Ç ActionMessage (proto)")
    print("3. –§–æ—Ä–º–∞—Ç StreamResponse")
    print("4. –§–æ—Ä–º–∞—Ç __MCP__ text_chunk")
    
    results = []
    
    # –¢–µ—Å—Ç 1: command_payload
    results.append(("–§–æ—Ä–º–∞—Ç command_payload", test_command_payload_format()))
    
    # –¢–µ—Å—Ç 2: ActionMessage
    results.append(("–§–æ—Ä–º–∞—Ç ActionMessage", test_action_message_proto()))
    
    # –¢–µ—Å—Ç 3: StreamResponse
    results.append(("–§–æ—Ä–º–∞—Ç StreamResponse", test_stream_response_format()))
    
    # –¢–µ—Å—Ç 4: __MCP__ text_chunk
    results.append(("–§–æ—Ä–º–∞—Ç __MCP__ text_chunk", test_mcp_text_chunk_format()))
    
    # –ò—Ç–æ–≥–∏
    print("\n" + "="*80)
    print("–ò–¢–û–ì–ò –¢–ï–°–¢–ò–†–û–í–ê–ù–ò–Ø gRPC –§–û–†–ú–ê–¢–ê")
    print("="*80)
    
    passed = sum(1 for _, result in results if result)
    total = len(results)
    
    for name, result in results:
        status = "‚úÖ PASS" if result else "‚ùå FAIL"
        print(f"{status} - {name}")
    
    print(f"\nüìä –†–µ–∑—É–ª—å—Ç–∞—Ç: {passed}/{total} —Ç–µ—Å—Ç–æ–≤ –ø—Ä–æ–π–¥–µ–Ω–æ")
    
    if passed == total:
        print("\nüéâ –í—Å–µ —Ç–µ—Å—Ç—ã –ø—Ä–æ–π–¥–µ–Ω—ã! –§–æ—Ä–º–∞—Ç –¥–∞–Ω–Ω—ã—Ö –¥–ª—è gRPC –∫–æ—Ä—Ä–µ–∫—Ç–µ–Ω.")
        print("\n‚úÖ –ì–æ—Ç–æ–≤–æ –∫ –æ—Ç–ø—Ä–∞–≤–∫–µ –∫–ª–∏–µ–Ω—Ç—É:")
        print("   - command_payload –ø—Ä–∞–≤–∏–ª—å–Ω–æ —Å–µ—Ä–∏–∞–ª–∏–∑—É–µ—Ç—Å—è")
        print("   - __MCP__ –ø—Ä–µ—Ñ–∏–∫—Å –¥–æ–±–∞–≤–ª—è–µ—Ç—Å—è")
        print("   - StreamResponse –ø—Ä–∞–≤–∏–ª—å–Ω–æ —Ñ–æ—Ä–º–∏—Ä—É–µ—Ç—Å—è")
        print("   - ActionMessage –ø–æ–¥–¥–µ—Ä–∂–∏–≤–∞–µ—Ç—Å—è (–µ—Å–ª–∏ –∏—Å–ø–æ–ª—å–∑—É–µ—Ç—Å—è)")
        return 0
    else:
        print(f"\n‚ö†Ô∏è  {total - passed} —Ç–µ—Å—Ç(–æ–≤) –Ω–µ –ø—Ä–æ–π–¥–µ–Ω–æ.")
        return 1


if __name__ == "__main__":
    sys.exit(main())





